python-threatexchange/threatexchange/fetcher/fetch_api.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+"""
+The fetcher is the component that talks to external APIs to get and put signals
+
+@see Fetcher
+"""
+
+
+from dataclasses import dataclass
+from enum import Enum
+import typing as t
+
+from threatexchange.collab_config import CollaborationConfig
+from threatexchange.signal_type.signal_base import SignalType
+from . import fetch_state_base as state
+
+
+class SignalExchangeAPI:
+    """
+    An external APIs to get and maybe put signals
+
+    Fetchers ideally can checkpoint their progress, so that they can tail
+    updates from the server.
+
+    Is assumed that fetched signals have some metadata attached to them,
+    which is unique to that API. Additionally, it a assumed that there
+    might be multiple contributors to signals inside of an API.
+
+    The Fetcher can assume it is stateful, and
+    """
+
+    @classmethod
+    def name(cls) -> str:
+        """A human-friendly name unique to all fetchers"""
+        return cls.__name__
+
+    def resolve_owner(self, id: int) -> str:
+        """Convert an owner ID into an owner record"""
+        return ""
+
+    def fetch_once(
+        self, collab: CollaborationConfig, checkpoint: t.Any
+    ) -> state.FetchDelta:
+        """
+        Call out to external resources, pulling down one "batch" of content.
+        """
+        raise NotImplementedError
+
+    def report_seen(
+        self, s_type: SignalType, signal: str, metadata: state.FetchMetadata
+    ) -> None:
+        """
+        Report that this signal is a true positive
+        """
+        raise NotImplementedError
+
+    def report_opinion(
+        self, s_type: t.Type[SignalType], signal: str, opinion: state.SignalOpinion
+    ) -> None:
+        """
+        Weigh in on a signal for this collaboration
+        """
+        raise NotImplementedError
+
+    def report_true_positive(
+        self, s_type: t.Type[SignalType], signal: str, metadata: state.FetchMetadata
+    ) -> None:
+        """Report that a previously seen signal was a true positive"""
+        raise NotImplementedError
+
+    def report_false_positive(
+        self, s_type: t.Type[SignalType], signal: str, metadata: state.FetchMetadata
```

Comment:
Note that these don't need CollabConfig because FetchMetadata is unique to collab, and so should have enough context to make a write (TBD when actually trying to implement it though)